### PR TITLE
*: refactor index, use TableInfo and IndexInfo to create an Index

### PR DESCRIPTION
**Pull request by coocood**
Prepare for supporting prefixed index and case insensitive index.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @zimulala PTAL

File: executor/executor.go [hunk 1]
```diff
@@ -524,7 +524,7 @@ func (e *IndexRangeExec) Close() error {
 type IndexScanExec struct {
 	tbl         table.Table
 	tableAsName *model.CIStr
-	idx         *table.IndexedColumn
```
Comment (shenli) [inline executor/executor.go]: Need to update XIndexScanExec?

**Comment by shenli** [comment]
LGTM
Please move tablecodec out of xapi in the future PR.

File: executor/executor.go (hunk 1 above)
Comment (coocood) [inline executor/executor.go]: No, because `*tipb.IndexInfo` has complete column info.

File: table/tables/index.go [hunk 2]
```diff
@@ -92,36 +94,30 @@ func (c *indexIter) Next() (val []types.Datum, h int64, err error) {
...
-		prefix:    GenIndexPrefix(indexPrefix, indexID),
+		tblInfo: tableInfo,
+		idxInfo: indexInfo,
+		prefix:  kv.Key(tablecodec.EncodeTableIndexPrefix(tableInfo.ID, indexInfo.ID)),
 	}
-
 	return index
 }
 
+// Meta return index info.
```
Comment (zimulala) [inline table/tables/index.go]: return -> returns

File: table/table.go [hunk 3]
```diff
@@ -36,8 +36,9 @@ var (
 	errDuplicateColumn = terror.ClassTable.New(codeDuplicateColumn, "duplicate column")
 
 	errGetDefaultFailed = terror.ClassTable.New(codeGetDefaultFailed, "get default value fail")
-	errIndexOutBound    = terror.ClassTable.New(codeIndexOutBound, "index column offset out of bound")
 
+	// ErrIndexOutBound returns for index column offset out of bound.
+	ErrIndexOutBound = terror.ClassTable.New(codeIndexOutBound, "index column offset out of bound")
```
Comment (zimulala) [inline table/table.go]: Add error code.

Comment (zimulala) [inline table/table.go]: Ignore it.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**